### PR TITLE
[DF] Throw if thread pool size changed after RDF construction

### PR DESCRIPTION
**Pull request by eguiraud**
...and add a test. This fixes ROOT-10346.

### CODE REVIEW

File: tree/dataframe/src/RLoopManager.cxx [hunk 1]
```diff
@@ -541,10 +542,25 @@ unsigned int RLoopManager::GetNextID()
    return id;
 }
 
+namespace {
+static void CheckIfPoolSizeChanged(unsigned int nSlots)
+{
+   const auto poolSize = ROOT::GetImplicitMTPoolSize();
+   if (poolSize > 0 && poolSize != nSlots) {
+      Warning("RLoopManager::Run",
+              "When the RDataFrame was constructed the size of the thread pool was %d but when starting the event loop "
```
Comment (hageboeck) [inline tree/dataframe/src/RLoopManager.cxx:551]:
```suggestion
              "When the RDataFrame was constructed, the size of the thread pool was %d, but when starting the event loop, "
```

File: tree/dataframe/test/dataframe_concurrency.cxx [hunk 2]
```diff
@@ -41,4 +41,18 @@ TEST(RDFConcurrency, NestedParallelismBetweenDefineCalls)
...
+TEST(RDFSimpleTests, WarnOnPoolSizeMismatch)
+{
+   ROOT::RDataFrame df(1);
+   ROOT::EnableImplicitMT(3);
+   testing::internal::CaptureStderr();
+   df.Count().GetValue();
+   const std::string err = testing::internal::GetCapturedStderr();
+   EXPECT_EQ(err, "Warning in <RLoopManager::Run>: When the RDataFrame was constructed the size of the thread pool was "
+                  "1 but when starting the event loop it was 3. Make sure to call ROOT::EnableImplicitMT before "
+                  "constructing the RDataFrame.\n");
```
Comment (hageboeck) [inline tree/dataframe/test/dataframe_concurrency.cxx:56]:
```suggestion
   EXPECT_EQ(err, "Warning in <RLoopManager::Run>: When the RDataFrame was constructed, the size of the thread pool was "
                  "1, but when starting the event loop, it was 3. Make sure to call ROOT::EnableImplicitMT before "
                  "constructing the RDataFrame.\n");
```

File: tree/dataframe/src/RLoopManager.cxx (hunk 1 above)
Comment (eguiraud) [reply tree/dataframe/src/RLoopManager.cxx]: there was one extra comma in the suggestion, added a commit with the fix

**Comment by eguiraud** [comment]
@hageboeck the number of tasks is fixed, it depends on the number of clusters in the dataset. If the thread-pool size during the event loop is smaller than the number of slots RDF prepared, a bit of space and time is wasted but no harm is done, the extra slots are unused. If the thread-pool size during the event loop is larger than the number of slots RDF prepared, builds with asserts will fire an assertion, builds without asserts will incur in undefined behavior (most likely a crash: `pop_back` will be called on an empty `std::deque`).

File: tree/dataframe/src/RLoopManager.cxx (hunk 1 above)
Comment (hageboeck) [reply tree/dataframe/src/RLoopManager.cxx]: There were three extra commas, but people will also understand without them. :-)

**Comment by hageboeck** [comment]
Ok, at least people see a warning first and **then** the crash.

**Comment by eguiraud** [comment]
@hageboeck I don't know, we could throw an exception instead, what do you think? I just went with what was requested in https://sft.its.cern.ch/jira/browse/ROOT-10346

**Comment by hageboeck** [comment]
If you know that you might crash or yield undefined results, an exception is probably better. Users won't have to look through 16 stack traces with weird stuff.
If you can continue running, but it's suboptimal, warning and keeping running is probably fine. So maybe you implement the `>` check and throw?

**Comment by eguiraud** [comment]
@hageboeck how about now?

**Pull request merged**